### PR TITLE
DOC: tweak wording on Figure.show warning

### LINKED ISSUE #22169: [Doc]: figure.show works beyond what is documented
### Documentation Link

https://matplotlib.org/devdocs/api/figure_api.html?highlight=figure%20show#matplotlib.figure.Figure.show

### Problem

From the linked documentation

> Proper use cases for Figure.show include running this from a GUI application or an IPython shell.
> 
> If you're running a pure python shell or executing a non-GUI python script, you should use `matplotlib.pyplot.show` instead, which takes care of managing the event loop for you.

I have just checked, and `figure.show()` works ALSO in a pure python shell  (here it's 3.9, in Suse Tumbleweed).

### Suggested improvement

Specify that after version 3.x (??) of the interpreter the event loop is managed also by a vanilla Python REPL.

### Matplotlib Version

3.5.1

### PR DESCRIPTION
**Pull request by tacaswell**
## PR Summary

closes #22169

## PR Checklist

**Documentation and Tests**
- [x] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).

### CODE REVIEW

**Review by timhoffm — APPROVED** [review]

File: lib/matplotlib/figure.py [hunk 1]
```diff
@@ -2614,12 +2614,16 @@ def show(self, warn=True):
...
-            python script, you should use `matplotlib.pyplot.show` instead,
-            which takes care of managing the event loop for you.
+            Proper use cases for `.Figure.show` include running this from a GUI
+            application (where there is persistently an event loop running) or
+            from a shell with in input hook configured (IPython will do this).
+            Some, but not all, GUI toolkits will register an input hook on
+            import.  See :ref:`cp_integration` for more details.
+
+            If you're in a shell without input hook integration or executing a
+            python script, you should use `matplotlib.pyplot.show(block=True)`
```
Comment (timhoffm) [inline lib/matplotlib/figure.py:2624]: This link does not work either make it literal or 
```suggestion
            python script, you should use `matplotlib.pyplot.show` with ``block=True``
```

File: lib/matplotlib/figure.py [hunk 2]
```diff
@@ -2614,12 +2614,16 @@ def show(self, warn=True):
...
-            GUI application or an IPython shell.
-
-            If you're running a pure python shell or executing a non-GUI
-            python script, you should use `matplotlib.pyplot.show` instead,
-            which takes care of managing the event loop for you.
+            Proper use cases for `.Figure.show` include running this from a GUI
+            application (where there is persistently an event loop running) or
+            from a shell with in input hook configured (IPython will do this).
+            Some, but not all, GUI toolkits will register an input hook on
+            import.  See :ref:`cp_integration` for more details.
```
Comment (jklymak) [inline lib/matplotlib/figure.py:2621]: My apologies, but I find these changes much more confusing than the original.  What is an "input hook"?  What is a "GUI toolkit", and why do they register an "input hook"?  Where do they register it? and what does "on import" mean?  On import of what?

Comment (jklymak) [reply lib/matplotlib/figure.py:2621]: Note that there is a typo in the previous paragraph "an GUI even loop"...

File: lib/matplotlib/figure.py [hunk 3]
```diff
@@ -2614,12 +2614,16 @@ def show(self, warn=True):
...
+
+            If you're in a shell without input hook integration or executing a
+            python script, you should use `matplotlib.pyplot.show` with
+            ``block=True`` instead, which takes care of starting and running
```
Comment (jklymak) [inline lib/matplotlib/figure.py:2625]: block=True is the default?

Comment (tacaswell) [reply lib/matplotlib/figure.py:2625]: The default depends on if we are in "interactive mode" or not.

File: lib/matplotlib/figure.py (hunk 2 above)
Comment (tacaswell) [reply lib/matplotlib/figure.py:2621]: It does include a link to the (much) longer explanation. 

#22169 noted that they found a case where it _did_ work.  The goal was to add a few more details about when you should expect this to work or not.

File: lib/matplotlib/figure.py [hunk 4]
```diff
@@ -2614,12 +2614,16 @@ def show(self, warn=True):
...
-            GUI application or an IPython shell.
-
-            If you're running a pure python shell or executing a non-GUI
-            python script, you should use `matplotlib.pyplot.show` instead,
-            which takes care of managing the event loop for you.
+            Proper use cases for `.Figure.show` include running this from a GUI
+            application (where there is persistently an event loop running) or
+            from a shell with an input hook configured (IPython will do this).
+            Some, but not all, GUI toolkits will register an input hook on
+            import.  See :ref:`cp_integration` for more details.
```
Comment (jklymak) [inline lib/matplotlib/figure.py:2621]: I think being specific helps:

```suggestion
            Use cases for `.Figure.show` include running this from a GUI
            application (where there is persistently an event loop running) or
            from a shell, like IPython, that install an input hook to allow the interactive 
            shell to accept input while the figure is also being shown and interactive.
            Some, but not all, GUI toolkits will register an input hook on
            import.  See :ref:`cp_integration` for more details.
```

However, I'm still not sure what you mean by a "GUI toolkit" in this context.  Do you mean our backends?  Or do you mean Qt etc install an input hook on import, even in the absence of Matplotlib?

Comment (tacaswell) [reply lib/matplotlib/figure.py]: Thank you, that is much better!

**Review by dstansby — APPROVED** [review]

**Comment by QuLogic** [comment]
meeseeksdev backport to v3.6.x

**Pull request merged**